### ISSUE TITLE
fix neg64 + neg32 + sar + rep movsb

### ISSUE THREAD

**Issue by brandonros** [sha0coder/scemu#15]
cleaning up from https://github.com/sha0coder/scemu/issues/14

```shell
 RUST_BACKTRACE=1 cargo run --release -- -6 -vvv -f /Users/brandonros/Downloads/DTS9_PatcherV.exe
1354 0x144fad176: ret ; ret-addr: 0x1a2aa2115 ret-value: 0x5db9bfa4 
```

let me see where that gets us on x64dbg

no way we can implement qemu testing of instructions easily + performantly, right?

[2022-09-18.csv](https://github.com/sha0coder/scemu/files/9594998/2022-09-18.csv)

```
	pre-registers: rax: 0x4dbdd49 rbx: 0x0 rcx: 0x704f43db rdx: 0x100000000 rsi: 0x22dfe8 rdi: 0x144e471df rbp: 0x144f06171 rsp: 0x22dde0 rip: 0x14501058e
1338 0x14501058e: xor   eax,596262EDh
	post-registers: rax: 0x5db9bfa4 rbx: 0x0 rcx: 0x704f43db rdx: 0x100000000 rsi: 0x22dfe8 rdi: 0x144e471df rbp: 0x144f06171 rsp: 0x22dde0 rip: 0x14501058e

```

it's kind of worse than that because by the time we get to `14501058E` `rax` is already wrong

I did very gross stuff (not PR worth) to your code where I talk pre-op reg + post-op reg and diff them

```
536 | 0000000144EE03AE | FFC8                     | dec eax                                 | rax: 5968D677-> 5968D676                          |                                                   |
```

eax is wrong by the time we are here

![image](https://user-images.githubusercontent.com/8949910/190926989-6c5d512d-7b82-4d6f-a608-008fe9c51b41.png)

this helps me track back against x64dbg trace

```
diff_reg: rip = 144f558ed rax ff98 -> ffffff98; good
diff_reg: rip = 144f558f5 rax ffffff98 -> 9b4a3d0; good
diff_reg: rip = 144f558fb rax 9b4a3d0 -> c43cb9ae; bad, should be F089AEA9
diff_reg: rip = 144f558fe rax c43cb9ae -> bafb4f23; 
diff_reg: rip = 144f55903 rax bafb4f23 -> 4504b0dd; 
diff_reg: rip = 144fcbe78 rax 4504b0dd -> 4504b0de; 
diff_reg: rip = 144fcbe7f rax 4504b0de -> 4dbdd4a; 
diff_reg: rip = 144ee03ae rax 4dbdd4a -> 4dbdd49; 
diff_reg: rip = 14501058e rax 4dbdd49 -> 5db9bfa4; 
```

so now problem with r11d? from 144F558FB

```
0000000144F558FB | 41:33C3                  | xor eax,r11d                            |
```

**Comment by sha0coder** [comment]
It's getting carry flag wrongly, rcl works ok but with wrong carry the result is not ok.
1302 0x144f0618a: rcl   rax,8Eh       0x38a91 -> 0xe2a46000 
--- console ---
=>f
--- flags ---
cf: true
pf: true

on the trace.csv:
0519,0000000144F0618A,48:C1D0 8E,"rcl rax,8E",rax: 38A91-> E2A44000,,

without the carry would match perfectly the result eax.

**Comment by sha0coder** [comment]
Regarding qemu, one option is doing inline assembly for testing.
For example a test mode when all the instructions are compared with its inline assembly.

fn shld(prams) {
emulated
}

fn cpu_shld(param) {
  asm("shld p1, p2")
}

it would need an unsafe block but only used on test mode.

**Comment by sha0coder** [comment]
This sets the carry flag but its ok, i checked the flags on x64dbg
1298 0x144f0617b: cmp   r12w,5347h

**Comment by sha0coder** [comment]
Regarding the xor, it seems ok to me:
1325 0x144f558fb: xor   eax,r11d
    r11d: 0xcd881a7e
    eax: 0x9b4a3d0 -> 0xc43cb9ae

>>> hex( 0x9b4a3d0 ^ 0xcd881a7e )
'0xc43cb9ae'

**Comment by sha0coder** [comment]
761 0x144f916ef: pop   r11 ;0x144f58011              r11 good
771 0x144ed423c: adc   al,r11b                             rax bad
783 0x144ed4269: xor   al,r11b
799 0x144edddbc: xor   r11b,al
809 0x144eddde3: xor   r10d,r11d
811 0x144eddde7: test  r11b,r9b
813 0x144eddded: test  r11w,997h
823 0x144fbed17: push  r11 ;0x144f58039          r11 bad should be 1BB09DE2F (because of rax)

**Comment by sha0coder** [comment]
I prepared --test mode with several inline asm

target/release/scemu -f shellcodes64/DTS9_PatcherV.exe -6 -vv --test
...
1252 0x144ecec31: neg   rcx
thread 'main' panicked at '0xfffffffff92d6bae should be 0x74256658f92d6bae', src/emu.rs:3746:33

Im detecting bugs in this way.

**Comment by sha0coder** [comment]
more than 98,000 instructions emulated, and arrive to possible api call.

target/release/scemu -f shellcodes64/DTS9_PatcherV.exe -6 -vv --test
...
98008 0x144f1ad9f: mov   rcx,rax
98009 0x144f1ada2: jmp   0000000144FE1F3Ah
98010 0x144fe1f3a: jmp   0000000144EF3191h
98011 0x144ef3191: sub   ebx,1
98012 0x144ef3197: jmp   0000000145003DEEh
98013 0x145003dee: jne   0000000144FDE4D5h not taken 
98014 0x145003df4: mov   rax,[rbp]
98015 0x145003df9: sub   rsp,20h
98016 0x145003e00: jmp   0000000144F93FB4h
98017 0x144f93fb4: call  rax
/!\ setting rip to non mapped addr 0x895e4b70

**Comment by sha0coder** [comment]
yes please

**Comment by brandonros** [comment]
[2022-09-19.csv](https://github.com/sha0coder/scemu/files/9599686/2022-09-19.csv)

**Comment by brandonros** [comment]
[2022-09-19.csv](https://github.com/sha0coder/scemu/files/9599712/2022-09-19.csv)

**Comment by brandonros** [comment]
x64dbg gets to 143871 instructions executed before this call, scemu gets to 98017

so we're potentially missing 45k instructions

**Comment by sha0coder** [comment]
There is a huge bucle based on ja, jne, jbe and je
target/release/scemu -f shellcodes64/DTS9_PatcherV.exe -6 --test -n -vv | grep j | grep -v jmp

I will re-check flags on this instructions but probably is a bad register, i also will add more inline asm tests.
There are other samples well emulated so flags and branches in general may be correct.

**Comment by brandonros** [comment]
what are your thoughts on https://github.com/sha0coder/scemu/pull/16? maybe put it behind a different special flag to not break existing trace_regs?

if we could make it spit out a similar format to x64dbg we could just diff x64dbg trace csv output to scemu trace csv output if you think we're only at the beginning of having to diff what is going wrong

**Comment by brandonros** [comment]
```
 {
  "x64dbgLine": {
    "Index": "020D2",
    "Address": "0000000144FAD15F",
    "Bytes": "F3:A4",
    "Disassembly": "rep movsb ",
    "Registers": "rcx: FF-> FE rsi: 14F291-> 14F292 rdi: 14F251-> 14F252",
    "Memory": "",
    "Comments": ""
  },
  "scemuLine": {
    "index": "8403",
    "address": "0x144fad161",
    "disassembly": "sbb   edi,685844F5h"
  }
}
```

i see what it is, x64dbg is counting `rep movsb` as multiple instructions. not sure if you want to make that change too, probably not

**Comment by sha0coder** [comment]
#16 It's nice, i can't merge it because it's closed,
one option is doing like:
inc eax ;  eax: 100h-> 101h

this require to add code to every instruction but it worth.

**Comment by sha0coder** [comment]
awesome!!
this is the logic of rep movsb:

```rust
Mnemonic::Movsb => { 
                        self.show_instruction(&self.colors.light_cyan, &ins);
                  
                        if self.cfg.is_64bits {      
                            if ins.has_rep_prefix() {
                                loop {                                            
                                    let val = self.maps.read_byte(self.regs.rsi).expect("cannot read memory");
                                    self.maps.write_byte(self.regs.rdi, val);
                          
                                    if !self.flags.f_df {
                                        self.regs.rsi += 1;                 
                                        self.regs.rdi += 1;
                                    } else {
                                        self.regs.rsi -= 1;          
                                        self.regs.rdi -= 1;
                                    } 
                                                                        
                                    self.regs.rcx -= 1;
                                    if self.regs.rcx == 0 { 
                                        break 
                                    }
                                }

                            } else {
```

**Comment by sha0coder** [comment]
>>i see what it is, x64dbg is counting rep movsb as multiple instructions. not sure if you want to make that change too, probably not

what is better? printing the instruction for each iteration? It would be many noise in the -vv

**Comment by brandonros** [comment]
printing the instruction for each iteration is ok with maybe a -trace flag or whatever (for super verbose)

as long as the "number of instructions executed" gets incremented to match x64dbg, it will help us build tools to diff the logs while we are in these early stages of tracking down bugs i think

**Comment by brandonros** [comment]
first we get `index` of instruction executed lined up, then we make sure `rip` is same, then we can validate each instruction is modifying registers and memory same way

easier said than done :)

just to be extra clear:

we still have problem with where `rbp` is getting wrong value on stack (instead of kernel32.LocalAlloc). do you want to focus on that first, or do you want to focus on getting `rep movsb` to produce logs more similar to x64dbg trace?

**Comment by sha0coder** [comment]
I would be quiclky doing  triple -vvv to print every iteration and increment the counter in any case.

**Comment by sha0coder** [comment]
bug hunted:
target/release/scemu -f shellcodes64/DTS9_PatcherV.exe -6 -vvv --test  
...
2132 0x144f4b87a: sar   ah,6Ch
thread 'main' panicked at '0x0 should be 0xff', src/emu.rs:4012:37

and verified with x64dbg that is bad, should be 0xff.

**Comment by sha0coder** [comment]
bug hunted:
target/release/scemu -f shellcodes64/DTS9_PatcherV.exe -6 -vvv --test  
...
2132 0x144f4b87a: sar   ah,6Ch
thread 'main' panicked at '0x0 should be 0xff', src/emu.rs:4012:37

and verified with x64dbg that is bad, should be 0xff.

**Comment by brandonros** [comment]
you and i are making very good team so far :)

**Comment by sha0coder** [comment]
indeed :)

**Comment by sha0coder** [comment]
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=78cf0ae9b0a2fac4c4880fbc1bc05186

I implement it following the manual but don't match with processor result.

**Comment by sha0coder** [comment]
ok I forgot to apply the count mask. 
0x1f and 0x3f for 64bits

**Comment by sha0coder** [comment]
SHL SHR SAL SAR are supervised and ok so far,

SHLD has undefined behaviour on a cpu if counter > size, and DTS9_PatcherV.exe triggers this case several times.
In this case emulator and cpu can differ, the manual says "undefined behaviour".

**Comment by brandonros** [comment]
let's move forward with https://github.com/sha0coder/scemu/pull/16 (please feel free to clean it up/change it behind a different flag/anything you'd like) and i'll write some node.js javascript to diff x64dbg "registers changed" to scemu "registers changed"

**Comment by sha0coder** [comment]
merged,
it would be nice if it's possible to diff the flags.

**Comment by brandonros** [comment]
Right away I am seeing a problem that our `rsp` does not match `x64dbg` 

Can we get some kind of override to set where rsp should be starting so we can match x64dbg? We need to get everything as close to 1:1 so we can make sense/find problems.

```
pub fn init_stack64(&mut self) {
        let stack = self.maps.get_mem("stack");

        self.regs.rsp = 0x000000000014F4A8;
        self.regs.rbp = 0x0000000000000000;
        stack.set_base(0x0000000140000000);
        stack.set_size(0x6000);

        //assert!(self.regs.rsp < self.regs.rbp);
        assert!(self.regs.rsp > stack.get_base());
        assert!(self.regs.rsp < stack.get_bottom());
        //assert!(self.regs.rbp > stack.get_base());
        //assert!(self.regs.rbp < stack.get_bottom());
        assert!(stack.inside(self.regs.rsp));
        //assert!(stack.inside(self.regs.rbp));
    }
```

```
$ RUST_BACKTRACE=1 cargo run --target x86_64-apple-darwin --release -- -6 -vvv --regs -f /Users/brandonros/Downloads/DTS9_PatcherV.exe
   Compiling scemu v0.4.5 (/Users/brandonros/Desktop/scemu)
    Finished release [optimized] target(s) in 6.85s
     Running `target/x86_64-apple-darwin/release/scemu -6 -vvv --regs -f /Users/brandonros/Downloads/DTS9_PatcherV.exe`
initializing regs
loading memory maps
thread 'main' panicked at 'assertion failed: self.regs.rsp > stack.get_base()', src/emu.rs:190:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:48:5
   3: scemu::emu::Emu::init
   4: scemu::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

```

if you can help me get that working?

**Comment by brandonros** [comment]
![image](https://user-images.githubusercontent.com/8949910/191333583-d6aa8d9d-a035-4dd1-b94e-a937e5533e60.png)
not sure where stack is falling memory map wise

**Comment by brandonros** [comment]
```
	diff_reg: rip = 1448a76a4 rsp 14f4a8 -> 14f4a0; 
	diff_reg: rip = 144f84070 rsp 14f498 -> 14f490; 
	diff_reg: rip = 144f84075 r9 0 -> ac; 
	diff_reg: rip = 144f8407a rsp 14f490 -> 14f488; 
```

it is skipping `00000001448A76A9`?

**Comment by brandonros** [comment]
it's something about how jmp/call do a `break` in the match ins.mnemonic() running loop vs decoding loop, not sure if you have a good clean solution for that

**Comment by sha0coder** [comment]
if an eip/rip change happen has to break the decoder.iter()  if don't break is going to parse next instruction.
set_eip and set_rip in some cases enable the second break: self.force_break, I have to re-check those cases.
the second break reloads the code block.

**Comment by sha0coder** [comment]
the break dont break the match, break the loop.
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=22b814cca35da2a28f27b5fecf42400b

**Comment by brandonros** [comment]
can you clean it up/refactor it so i can get a reliable pre/post hook for register diffing so i can build per instruction comparison to x64dbg please? any ideas?

**Comment by brandonros** [comment]
i would love that/i think that would be perfect.

you will complete it maybe 1-2 days from now, or need more time?

**Comment by sha0coder** [comment]
the return is more confusing that the break in rust:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a65015c3fefbfe0709eebb55d3a84c61

in some way returns the match and in other way returns the function.

**Comment by sha0coder** [comment]
I put force_reload and force_break,
on force_break needs a reload but next instruction is next. ie: polymorphic code change
on force_reload the next instruction is other point.  ie: set_eip / set_rip

```rust
           if self.force_reload {                                  
                    self.force_reload = false;                       
                    break;                                    
                }                                                             

                if self.cfg.is_64bits {
                    self.regs.rip += sz as u64;                
                } else {          
                    self.regs.set_eip(self.regs.get_eip() + sz as u64);          
                }                                                                                 
                                                       
                if self.force_break {
                    self.force_break = false;
                    break;
                }
```

and the breaks because a parameter is bad  is going to be panic!()

**Comment by brandonros** [comment]
> I put force_reload and force_break, on force_break needs a reload but next instruction is next. ie: polymorphic code change on force_reload the next instruction is other point. ie: set_eip / set_rip
> 
> ```rust
>            if self.force_reload {                                  
>                     self.force_reload = false;                       
>                     break;                                    
>                 }                                                             
> 
>                 if self.cfg.is_64bits {
>                     self.regs.rip += sz as u64;                
>                 } else {          
>                     self.regs.set_eip(self.regs.get_eip() + sz as u64);          
>                 }                                                                                 
>                                                        
>                 if self.force_break {
>                     self.force_break = false;
>                     break;
>                 }
> ```
> 
> and the breaks because a parameter is bad is going to be panic!()

a lot better, can't wait to test it

if you could move one of the loops out into a different function or something, the nested loop was hard to follow code wise.

**Comment by sha0coder** [comment]
emu.step() is ready on the lib  https://crates.io/crates/libscemu

but with a rep will do the full rep for now,  
I have an idea of how to do the one-step-rep.

**Comment by sha0coder** [comment]
after an instruction is emulated, the rip is incremented in the size of the instruction.
on jmp/calls never has to to increment the rip. 
force_break does the rip += size
force_reload not, because rip has been alredy fixed

**Comment by sha0coder** [comment]
ok emu.step()  in case of rep does only one step, I haven't tested it well tho.

**Comment by brandonros** [comment]
do you have plans to pull libscemu into scemu?
